Unable to run dspace cleanup

## Describe the bug
When running `dspace cleanup -v` I get following error, on DSpace 7.6.2.

## To Reproduce
Steps to reproduce the behavior:
1. `dspace cleanup -v`

```
Found 77746 deleted bistream to cleanup
 - Deleting bitstream information (ID: ffffd693-8181-4c49-8031-0677d7205a26)
 - Deleting bitstream record from database (ID: ffffd693-8181-4c49-8031-0677d7205a26)
 - Deleting bitstream information (ID: fffdf265-2566-426f-8ba8-844623a7065b)
 - Deleting bitstream record from database (ID: fffdf265-2566-426f-8ba8-844623a7065b)
 - Deleting bitstream information (ID: fffd0b11-1935-43c4-98a2-e5add849ccbd)
 - Deleting bitstream record from database (ID: fffd0b11-1935-43c4-98a2-e5add849ccbd)
 - Deleting bitstream information (ID: fffa47ac-9da7-4144-a8c5-1c2ed67c3ec8)
 - Deleting bitstream record from database (ID: fffa47ac-9da7-4144-a8c5-1c2ed67c3ec8)
Exception: org.hibernate.exception.ConstraintViolationException: could not execute statement
javax.persistence.PersistenceException: org.hibernate.exception.ConstraintViolationException: could not execute statement
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1626)
	at org.hibernate.query.Query.getResultList(Query.java:165)
	at org.hibernate.query.criteria.internal.compile.CriteriaQueryTypeQueryAdapter.getResultList(CriteriaQueryTypeQueryAdapter.java:76)
	at org.dspace.core.AbstractHibernateDAO.executeCriteriaQuery(AbstractHibernateDAO.java:427)
	at org.dspace.core.AbstractHibernateDAO.list(AbstractHibernateDAO.java:162)
	at org.dspace.content.dao.impl.BitstreamDAOImpl.findDuplicateInternalIdentifier(BitstreamDAOImpl.java:68)
	at org.dspace.content.BitstreamServiceImpl.findDuplicateInternalIdentifier(BitstreamServiceImpl.java:356)
	at org.dspace.storage.bitstore.BitstreamStorageServiceImpl.cleanup(BitstreamStorageServiceImpl.java:294)
	at org.dspace.storage.bitstore.Cleanup.main(Cleanup.java:81)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:283)
	at org.dspace.app.launcher.ScriptLauncher.handleScript(ScriptLauncher.java:134)
	at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:99)
Caused by: org.hibernate.exception.ConstraintViolationException: could not execute statement
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:109)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:200)
	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3702)
	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3987)
	at org.hibernate.action.internal.EntityDeleteAction.execute(EntityDeleteAction.java:123)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:344)
	at org.hibernate.event.internal.DefaultAutoFlushEventListener.onAutoFlush(DefaultAutoFlushEventListener.java:57)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
	at org.hibernate.internal.SessionImpl.autoFlushIfRequired(SessionImpl.java:1372)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1452)
	at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1649)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1617)
	... 15 more
Caused by: org.postgresql.util.PSQLException: ERROR: update or delete on table "bitstream" violates foreign key constraint "bundle2bitstream_bitstream_id_fkey" on table "bundle2bitstream"
  Detail: Key (uuid)=(fffa47ac-9da7-4144-a8c5-1c2ed67c3ec8) is still referenced from table "bundle2bitstream".
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2725)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2412)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:371)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:502)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:419)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:194)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:155)
	at org.apache.commons.dbcp2.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:137)
	at org.apache.commons.dbcp2.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:137)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)
	... 29 more

```

## Expected behavior
Clean up database

--- a/dspace-api/src/main/java/org/dspace/content/BitstreamServiceImpl.java
+++ b/dspace-api/src/main/java/org/dspace/content/BitstreamServiceImpl.java
@@ -515,6 +515,28 @@ public class BitstreamServiceImpl extends DSpaceObjectServiceImpl<Bitstream> imp
             throw new IllegalStateException("Bitstream " + bitstream.getID().toString()
                     + " must be deleted before it can be removed from the database.");
         }
+
+        // Defensively remove any remaining bundle2bitstream references.
+        // Normally delete() already cleans these up, but orphaned rows from
+        // historical bugs can cause FK constraint violations on hard-delete.
+        final List<Bundle> bundles = bitstream.getBundles();
+        for (Bundle bundle : bundles) {
+            if (bitstream.equals(bundle.getPrimaryBitstream())) {
+                bundle.unsetPrimaryBitstreamID();
+            }
+            bundle.removeBitstream(bitstream);
+        }
+        bundles.clear();
+
+        // Remove any orphaned request items referencing this bitstream
+        Iterator<RequestItem> requestItems = requestItemService.findByBitstreamId(context, bitstream.getID());
+        while (requestItems.hasNext()) {
+            requestItemService.delete(context, requestItems.next());
+        }
+
+        // Remove any remaining authorization policies
+        authorizeService.removeAllPolicies(context, bitstream);
+
         bitstreamDAO.delete(context, bitstream);
     }
 

--- a/dspace-api/src/main/java/org/dspace/storage/bitstore/BitstreamStorageServiceImpl.java
+++ b/dspace-api/src/main/java/org/dspace/storage/bitstore/BitstreamStorageServiceImpl.java
@@ -280,6 +280,13 @@ public class BitstreamStorageServiceImpl implements BitstreamStorageService, Ini
                         continue;
                     }
 
+                    // Check whether the bitstore file should be kept before
+                    // expunging the database record, because expunge() would
+                    // make the bitstream entity stale for subsequent queries.
+                    boolean isRegistered = isRegisteredBitstream(bitstream.getInternalId());
+                    boolean hasDuplicate = !bitstreamService
+                        .findDuplicateInternalIdentifier(context, bitstream).isEmpty();
+
                     if (deleteDbRecords) {
                         log.debug("deleting db record");
                         if (verbose) {
@@ -292,16 +299,15 @@ public class BitstreamStorageServiceImpl implements BitstreamStorageService, Ini
                         bitstreamService.expunge(context, bitstream);
                     }
 
-                    if (isRegisteredBitstream(bitstream.getInternalId())) {
+                    if (isRegistered) {
                         context.uncacheEntity(bitstream);
-                        continue; // do not delete registered bitstreams
+                        continue; // do not delete registered bitstreams from the bitstore
                     }
 
-
-                    // Since versioning allows for multiple bitstreams, check if the internal
-                    // identifier isn't used on
-                    // another place
-                    if (bitstreamService.findDuplicateInternalIdentifier(context, bitstream).isEmpty()) {
+                    // Since versioning allows for multiple bitstreams, only
+                    // remove the file if no other bitstream shares this
+                    // internal identifier
+                    if (!hasDuplicate) {
                         this.getStore(bitstream.getStoreNumber()).remove(bitstream);
 
                         String message = ("Deleted bitstreamID " + bid + ", internalID " + bitstream.getInternalId());

--- a/dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V8.4_2026.03.28__Fix-bundle2bitstream-with-deleted-bitstreams.sql
+++ b/dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V8.4_2026.03.28__Fix-bundle2bitstream-with-deleted-bitstreams.sql
@@ -1,0 +1,17 @@
+--
+-- The contents of this file are subject to the license and copyright
+-- detailed in the LICENSE and NOTICE files at the root of the source
+-- tree and available online at
+--
+-- http://www.dspace.org/license/
+--
+
+-- Remove orphaned bundle2bitstream rows that reference deleted bitstreams.
+-- These orphaned rows prevent 'dspace cleanup' from expunging deleted
+-- bitstreams due to FK constraint violations.
+DELETE
+FROM bundle2bitstream
+WHERE bitstream_id IN
+      (SELECT bs.uuid
+       FROM bitstream AS bs
+       WHERE bs.deleted IS TRUE)

--- a/dspace-api/src/test/java/org/dspace/content/BitstreamExpungeIT.java
+++ b/dspace-api/src/test/java/org/dspace/content/BitstreamExpungeIT.java
@@ -1,0 +1,247 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.content;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.UUID;
+
+import org.apache.commons.io.IOUtils;
+import org.dspace.AbstractIntegrationTestWithDatabase;
+import org.dspace.app.requestitem.RequestItem;
+import org.dspace.app.requestitem.factory.RequestItemServiceFactory;
+import org.dspace.app.requestitem.service.RequestItemService;
+import org.dspace.authorize.factory.AuthorizeServiceFactory;
+import org.dspace.authorize.service.AuthorizeService;
+import org.dspace.builder.BitstreamBuilder;
+import org.dspace.builder.CollectionBuilder;
+import org.dspace.builder.CommunityBuilder;
+import org.dspace.builder.ItemBuilder;
+import org.dspace.builder.RequestItemBuilder;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.content.service.BitstreamService;
+import org.dspace.content.service.BundleService;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Integration tests for {@link BitstreamServiceImpl#expunge}.
+ *
+ * Verifies that expunge() defensively cleans up bundle2bitstream, requestitem,
+ * and resourcepolicy references before the hard-delete, so historical orphaned
+ * rows do not cause FK constraint violations.
+ *
+ * @author Bram Luyten (bram at atmire.com)
+ */
+public class BitstreamExpungeIT extends AbstractIntegrationTestWithDatabase {
+
+    private final BitstreamService bitstreamService =
+        ContentServiceFactory.getInstance().getBitstreamService();
+    private final BundleService bundleService =
+        ContentServiceFactory.getInstance().getBundleService();
+    private final RequestItemService requestItemService =
+        RequestItemServiceFactory.getInstance().getRequestItemService();
+    private final AuthorizeService authorizeService =
+        AuthorizeServiceFactory.getInstance().getAuthorizeService();
+
+    private Collection collection;
+
+    @Before
+    public void setup() {
+        context.turnOffAuthorisationSystem();
+        parentCommunity = CommunityBuilder.createCommunity(context).build();
+        collection = CollectionBuilder.createCollection(context, parentCommunity).build();
+        context.restoreAuthSystemState();
+    }
+
+    /**
+     * Smoke test the normal path: a bitstream that has been soft-deleted via
+     * delete() can be expunged. delete() already cleans up FK references, so
+     * expunge() should succeed without any defensive work.
+     */
+    @Test
+    public void testExpungeAfterNormalDelete() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        Item item = ItemBuilder.createItem(context, collection).withTitle("Test item").build();
+        Bitstream bitstream = BitstreamBuilder
+            .createBitstream(context, item, toInputStream("test content"))
+            .build();
+        UUID bitstreamId = bitstream.getID();
+
+        bitstreamService.delete(context, bitstream);
+        context.commit();
+
+        bitstream = bitstreamService.find(context, bitstreamId);
+        assertTrue("Bitstream should be marked as deleted", bitstream.isDeleted());
+
+        bitstreamService.expunge(context, bitstream);
+        context.commit();
+
+        assertNull("Bitstream should not exist after expunge",
+            bitstreamService.find(context, bitstreamId));
+
+        context.restoreAuthSystemState();
+    }
+
+    /**
+     * Simulate the historical bug scenario: a bitstream is marked deleted but
+     * its bundle2bitstream row was never cleaned up. expunge() must remove the
+     * leftover bundle association so the hard-delete does not hit a FK
+     * constraint violation.
+     */
+    @Test
+    public void testExpungeOrphanedBundleReference() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        Item item = ItemBuilder.createItem(context, collection).withTitle("Test item").build();
+        Bitstream bitstream = BitstreamBuilder
+            .createBitstream(context, item, toInputStream("test content"))
+            .build();
+        UUID bitstreamId = bitstream.getID();
+        UUID bundleId = bitstream.getBundles().get(0).getID();
+
+        // Mark the bitstream as deleted WITHOUT going through delete(), so the
+        // bundle relationship stays in place. This is the orphan state the PR
+        // is designed to recover from.
+        bitstream.setDeleted(true);
+        bitstreamService.update(context, bitstream);
+        context.commit();
+
+        bitstream = bitstreamService.find(context, bitstreamId);
+        assertEquals("Bundle association should still exist before expunge",
+            1, bitstream.getBundles().size());
+
+        bitstreamService.expunge(context, bitstream);
+        context.commit();
+
+        assertNull("Bitstream should be removed after expunge",
+            bitstreamService.find(context, bitstreamId));
+
+        Bundle bundle = bundleService.find(context, bundleId);
+        assertFalse("Bundle should no longer reference the bitstream",
+            bundle.getBitstreams().stream().anyMatch(b -> b.getID().equals(bitstreamId)));
+
+        context.restoreAuthSystemState();
+    }
+
+    /**
+     * Variant where the orphaned bitstream is also the bundle's primary
+     * bitstream. expunge() must unset the primary bitstream pointer before
+     * hard-deleting, otherwise the bundle_primary_bitstream_id_fkey constraint
+     * fires.
+     */
+    @Test
+    public void testExpungeOrphanedPrimaryBitstream() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        Item item = ItemBuilder.createItem(context, collection).withTitle("Test item").build();
+        Bitstream bitstream = BitstreamBuilder
+            .createBitstream(context, item, toInputStream("test content"))
+            .build();
+        UUID bitstreamId = bitstream.getID();
+        Bundle bundle = bitstream.getBundles().get(0);
+        UUID bundleId = bundle.getID();
+
+        bundle.setPrimaryBitstreamID(bitstream);
+        bundleService.update(context, bundle);
+
+        bitstream.setDeleted(true);
+        bitstreamService.update(context, bitstream);
+        context.commit();
+
+        // Re-fetch in a fresh session state, simulating the cleanup script
+        // path which iterates deleted bitstreams in a separate session.
+        bitstream = bitstreamService.find(context, bitstreamId);
+        bitstreamService.expunge(context, bitstream);
+        context.commit();
+
+        assertNull("Bitstream should be removed after expunge",
+            bitstreamService.find(context, bitstreamId));
+        assertNull("Bundle should no longer have a primary bitstream",
+            bundleService.find(context, bundleId).getPrimaryBitstream());
+
+        context.restoreAuthSystemState();
+    }
+
+    /**
+     * Variant where a RequestItem references the orphaned bitstream. expunge()
+     * must remove the request item so the requestitem_bitstream_id_fkey
+     * constraint does not fire on the hard-delete.
+     */
+    @Test
+    public void testExpungeOrphanedRequestItem() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        Item item = ItemBuilder.createItem(context, collection).withTitle("Test item").build();
+        Bitstream bitstream = BitstreamBuilder
+            .createBitstream(context, item, toInputStream("test content"))
+            .build();
+        UUID bitstreamId = bitstream.getID();
+
+        RequestItemBuilder.createRequestItem(context, item, bitstream).build();
+
+        bitstream.setDeleted(true);
+        bitstreamService.update(context, bitstream);
+        context.commit();
+
+        Iterator<RequestItem> before = requestItemService.findByBitstreamId(context, bitstreamId);
+        assertTrue("RequestItem should exist before expunge", before.hasNext());
+
+        bitstream = bitstreamService.find(context, bitstreamId);
+        bitstreamService.expunge(context, bitstream);
+        context.commit();
+
+        assertNull("Bitstream should be removed after expunge",
+            bitstreamService.find(context, bitstreamId));
+        assertFalse("RequestItem should be removed after expunge",
+            requestItemService.findByBitstreamId(context, bitstreamId).hasNext());
+
+        context.restoreAuthSystemState();
+    }
+
+    /**
+     * Verify that policies attached to the bitstream are removed by expunge(),
+     * so resource policy rows do not survive the hard-delete.
+     */
+    @Test
+    public void testExpungeRemovesPolicies() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        Item item = ItemBuilder.createItem(context, collection).withTitle("Test item").build();
+        Bitstream bitstream = BitstreamBuilder
+            .createBitstream(context, item, toInputStream("test content"))
+            .build();
+        UUID bitstreamId = bitstream.getID();
+
+        assertFalse("Bitstream should have inherited policies",
+            authorizeService.getPolicies(context, bitstream).isEmpty());
+
+        bitstream.setDeleted(true);
+        bitstreamService.update(context, bitstream);
+        context.commit();
+
+        bitstream = bitstreamService.find(context, bitstreamId);
+        bitstreamService.expunge(context, bitstream);
+        context.commit();
+
+        assertNull("Bitstream should be removed after expunge",
+            bitstreamService.find(context, bitstreamId));
+
+        context.restoreAuthSystemState();
+    }
+
+    private InputStream toInputStream(String content) {
+        return IOUtils.toInputStream(content, "UTF-8");
+    }
+}

